### PR TITLE
Add timeout for eth_getWork call

### LINKED ISSUE #1818: Add an option for mining work timeout.
When set, `eth_getWork` should produce an error if the work package was not changed for specified time.

### PR DESCRIPTION
**Pull request by nipunn1313**
Fixes #1818.
I implemented the timeout by comparing the current time to the block timestamp.
We obviously have to add some documentation as well, but I wanted to get this out for review first.
Also had to play some games with the tests to get it to work.

### CODE REVIEW

**Comment by rphmeier** [comment]
I think this also has to come with a command-line flag that is off by default (and off in `--geth` mode, until geth supports something similar) as this is a breaking change in the public API.

**Comment by gavofyork** [comment]
i think off in geth mode alone is sufficient; for 1.x our public API is only the RPCs covered under `--geth`. we can't be freezing RPCs under non-geth mode yet since they're still evolving. we'll declare non-geth mode RPCs (or a subset thereof) as public API and freeze them at 2.0.

**Comment by tomusdrw** [comment]
It's not a breaking change - it's optional parameter, look at the test cases:
https://github.com/ethcore/parity/pull/1975/files#diff-dfca3d175a69ed8008f6e6cd851722f6R837

File: rpc/src/v1/impls/eth.rs [hunk 1]
```diff
@@ -516,7 +517,7 @@ impl<C, S: ?Sized, M, EM> Eth for EthClient<C, S, M, EM> where
 
 	fn work(&self, params: Params) -> Result<Value, Error> {
 		try!(self.active());
-		try!(expect_no_params(params));
+		let (no_new_work_timeout,) = from_params::<(u64,)>(params).unwrap_or((0,));
```
Comment (nipunn1313) [inline rpc/src/v1/impls/eth.rs]: It's treated as on optional parameter which defaults to 0 here.

File: rpc/src/v1/tests/mocked/eth.rs [hunk 2]
```diff
@@ -852,3 +853,28 @@ fn should_not_return_block_number() {
...
+
+#[test]
+fn rpc_get_work_should_timeout() {
+	let eth_tester = EthTester::default();
+	eth_tester.miner.set_author(Address::from_str("d46e8dd67c5d32be8058bb8eb970870f07244567").unwrap());
+	eth_tester.client.set_latest_block_timestamp(get_time().sec as u64 - 1000);  // Set latest block to 1000 seconds ago
+	let hash = eth_tester.miner.map_sealing_work(&*eth_tester.client, |b| b.hash()).unwrap();
+
+	// Request with timeout of 0 seconds. This should work since we're disabling timeout.
+	let request = r#"{"jsonrpc": "2.0", "method": "eth_getWork", "params": [], "id": 1}"#;
```
Comment (nipunn1313) [inline rpc/src/v1/tests/mocked/eth.rs]: This test case shows that the API is backward compatible. I should probably also add a test case which explicitly passes in 0.

**Pull request merged**